src/core/experimental/refcount.d

```diff
@@ -0,0 +1,782 @@
+/**
+  This module provides a composable reference count implementation in the form
+  of `__RefCount`.
+*/
+module core.experimental.refcount;
+
+/**
+ * A qualified reference counted struct that is intended to be composed by user
+ * defined types that desire to implement manual memory management by means of
+ * reference counting. Note to user: The internal implementation uses malloc/free.
+ *
+ * `__RefCount` was designed to be composed as a field inside the user defined type.
+ * The user is responsible to initialize the `__RefCount` in the constructor of his
+ * type. The user will call the `isUnique()` method to decide if this is the last
+ * reference to his type so he can safely deallocate his own managed memory.
+ *
+ * `Important`: the `__RefCount` must be initialized through a call to the
+ * constructor before being used.
+ */
+struct __RefCount
+{
+    import core.atomic : atomicOp;
+
+    alias CounterType = uint;
+    private CounterType* rc = null;
+
+    /**
+     * Return a boolean value denoting if this can be used in a shared context.
+     *
+     * Returns:
+     *      `true` if this started off as an immutable object; `false` otherwise.
+     *
+     * Complexity:
+     *      $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe scope
+    bool isShared() const
+    {
+        // Faster than ((cast(size_t) rc) % 8) == 0;
+        return !((cast(size_t) rc) & 7);
+    }
+
+    /*
+     * Perform `rc op val` operation. Use atomics if this is shared.
+     *
+     * Returns:
+     *      The result of the operation.
+     */
+    @nogc nothrow pure @trusted scope
+    private CounterType rcOp(string op)(CounterType val) const
+    {
+        if (isShared())
+        {
+            return cast(CounterType)(atomicOp!op(*(cast(shared CounterType*) rc), val));
+        }
+        else
+        {
+            mixin("return cast(CounterType)(*(cast(CounterType*) rc)" ~ op ~ "val);");
+        }
+    }
+
+    /**
+     * Creates a new `__RefCount` instance. It's memory is internally managed with
+     * malloc/free.
+     *
+     * Params:
+     *      _ = an unused int value; required because structs don't have a
```

Comment:
Wouldn't a factory function be better?

Reply:
Something like `makeRefCount` and `makeSharedRefCount` ?

Reply:
Yes. In the unittest it looks weird to pass 1 to a constructor just to get around no-arg ctors.